File: RB101_Programming_Foundations/Small Probems/Medium1/rotation3.rb
# iterate each index of the number string array, rotate the current index to the right

require 'pry'

# def max_rotation(number)
#   digits = number.to_s.chars

#   digits.each_with_index do |_ , idx|
#     digits[idx..-1] = rotate_right(digits[idx..-1])
#     break if idx == digits.length - 2
#   end
#   digits.join.to_i
# end



# def rotate_right(digits)
#   digits[1..-1] + [digits[0]]
# end

def max_rotation(number)
  digits = number.to_s.chars

  digits.size.times do |idx|
    digit = digits.delete_at(idx)
    digits << digit
    break if idx >= digits.size - 2
  end

  digits.join.to_i

end

puts max_rotation(735291) == 321579
puts max_rotation(3) == 3
puts max_rotation(35) == 53
puts max_rotation(105) == 15 # the leading zero gets dropped
puts max_rotation(8_703_529_146) == 7_321_609_845
puts max_rotation(735000)

File: js210_fundamentals_of_JS_Programming/problem_sets/medium1/rotation3.js
function maxRotation(number) {
  let length = String(number).length;

  for (let index = length; index > 1; index -= 1) {
    number = rotateRightmostDigits(number, index);
  }
  return number;
}

function rotateRightmostDigits(number, rotateDigit) {
  let digits = String(number);
  let rotatedDigits = digits.slice(0, -rotateDigit).concat(rotateString(digits.slice(-rotateDigit)));
  return parseInt(rotatedDigits);
}

function rotateString(string) {
  return string.slice(1) + string[0];
}

console.log(maxRotation(735291)); // 321579
console.log(maxRotation(3)); // 3
console.log(maxRotation(35)); // 53
console.log(maxRotation(105)); // 15 -- the leading zero gets dropped
console.log(maxRotation(8703529146)); // 7321609845
